# DATA EXPLORATION USING PANDAS

In [1]:
import pandas as pd

In [2]:
sample=pd.read_csv('F://Abi//python//used_cars_data.csv')

###### About Dataset
Source: https://zenodo.org/record/6438518

Dataset that contains information on the different second-hand cars that were for sale on the www.flexicar.es website for the province of Barcelona as of April 8, 2022.

790 cars are included in the dataset, being the cars that were for sale on the Flexicar website for the province of Barcelona at the time of extraction, described by the following attributes:

In [3]:
sample

Unnamed: 0       brand     model  price (eur)  \
0             0        SEAT     Ibiza         8990   
1             1     Hyundai       i30         9990   
2             2         BMW   Serie 5        13490   
3             3  Volkswagen      Golf        24990   
4             4        Opel     Corsa        10460   
..          ...         ...       ...          ...   
787         787  Volkswagen      Golf        13990   
788         788         Kia  Sportage        24990   
789         789      Abarth       500        17990   
790         790  Volkswagen    Tiguan        14990   
791         791       Dacia    Dokker        14390   

                                       engine  year  mileage (kms)      fuel  \
0                       SC 1.2 TSI 90cv Style  2016          67000  Gasolina   
1                        1.6 CRDi 110cv Tecno  2014         104868    Diésel   
2                                530d Touring  2011         137566    Diésel   
3                   GTI 2.0 TSI 169kW (230CV)  2018          44495  Gasolina   
4                        1.4 Expression 90 CV  2016          69800  Gasolina   
..                                        ...   ...            ...       ...   
787                 Edition 1.6 TDI 110CV BMT  2016          84040    Diésel   
788            1.6 GDi 97kW (132CV) Basic 4x2  2018          65872  Gasolina   
789  1.4 16v T-Jet 595 118kW (160CV) Pista E6  2019          28830  Gasolina   
790  2.0 TDI 177cv DSG 4x4 Sport BMotion Tech  2014         162895    Diésel   
791            SL Trotamundos dCi 66kW (90CV)  2018           9800    Diésel   

        gearbox    location Unnamed: 10  
0        Manual  Granollers         NaN  
1        Manual  Viladecans         NaN  
2    Automatica  Viladecans              
3        Manual  Viladecans         NaN  
4        Manual  Sabadell 1         NaN  
..          ...         ...         ...  
787      Manual        Gavá         NaN  
788      Manual  Viladecans         NaN  
789      Manual      Mataró         NaN  
790  Automatica      Mataró         NaN  
791      Manual    Badalona         NaN  

[792 rows x 11 columns]

##### To veiw the dataset in python.

In [4]:
sample.head()

Unnamed: 0       brand    model  price (eur)                     engine  \
0           0        SEAT    Ibiza         8990      SC 1.2 TSI 90cv Style   
1           1     Hyundai      i30         9990       1.6 CRDi 110cv Tecno   
2           2         BMW  Serie 5        13490               530d Touring   
3           3  Volkswagen     Golf        24990  GTI 2.0 TSI 169kW (230CV)   
4           4        Opel    Corsa        10460       1.4 Expression 90 CV   

   year  mileage (kms)      fuel     gearbox    location Unnamed: 10  
0  2016          67000  Gasolina      Manual  Granollers         NaN  
1  2014         104868    Diésel      Manual  Viladecans         NaN  
2  2011         137566    Diésel  Automatica  Viladecans              
3  2018          44495  Gasolina      Manual  Viladecans         NaN  
4  2016          69800  Gasolina      Manual  Sabadell 1         NaN

##### The first five observation of the  data.

In [5]:
sample.tail()

Unnamed: 0       brand     model  price (eur)  \
787         787  Volkswagen      Golf        13990   
788         788         Kia  Sportage        24990   
789         789      Abarth       500        17990   
790         790  Volkswagen    Tiguan        14990   
791         791       Dacia    Dokker        14390   

                                       engine  year  mileage (kms)      fuel  \
787                 Edition 1.6 TDI 110CV BMT  2016          84040    Diésel   
788            1.6 GDi 97kW (132CV) Basic 4x2  2018          65872  Gasolina   
789  1.4 16v T-Jet 595 118kW (160CV) Pista E6  2019          28830  Gasolina   
790  2.0 TDI 177cv DSG 4x4 Sport BMotion Tech  2014         162895    Diésel   
791            SL Trotamundos dCi 66kW (90CV)  2018           9800    Diésel   

        gearbox    location Unnamed: 10  
787      Manual        Gavá         NaN  
788      Manual  Viladecans         NaN  
789      Manual      Mataró         NaN  
790  Automatica      Mataró         NaN  
791      Manual    Badalona         NaN

##### The last five observation of the data.

In [6]:
sample.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 792 entries, 0 to 791
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Unnamed: 0     792 non-null    int64 
 1   brand          792 non-null    object
 2   model          792 non-null    object
 3   price (eur)    792 non-null    int64 
 4   engine         792 non-null    object
 5   year           792 non-null    int64 
 6   mileage (kms)  792 non-null    int64 
 7   fuel           792 non-null    object
 8   gearbox        792 non-null    object
 9   location       792 non-null    object
 10  Unnamed: 10    1 non-null      object
dtypes: int64(4), object(7)
memory usage: 68.2+ KB


##### The datatype and memory usage of the data.

In [7]:
sample.describe()

Unnamed: 0   price (eur)         year  mileage (kms)
count  792.000000    792.000000   792.000000     792.000000
mean   395.500000  15973.383838  2016.404040   83228.833333
std    228.774999   6984.338370     2.452251   46202.118589
min      0.000000   4990.000000  2004.000000    1258.000000
25%    197.750000  11355.000000  2015.000000   47145.500000
50%    395.500000  14990.000000  2017.000000   80065.500000
75%    593.250000  18490.000000  2018.000000  116321.750000
max    791.000000  82580.000000  2021.000000  220000.000000

##### The statistical functions of the dataset.

In [8]:
sample.shape

(792, 11)

##### 792 Observations and 11 Varibles.

In [9]:
sample.columns

Index(['Unnamed: 0', 'brand', 'model', 'price (eur)', 'engine', 'year',
       'mileage (kms)', 'fuel', 'gearbox', 'location', 'Unnamed: 10'],
      dtype='object')

##### The column names of the dataset.

In [10]:
sample.index

RangeIndex(start=0, stop=792, step=1)

##### The range of the data from 0 to 792.

In [11]:
sample.sort_values(['price (eur)'],ascending=True)

Unnamed: 0     brand                     model  price (eur)  \
708         708      Ford                    Fiesta         4990   
779         779   Citroen                        C2         4990   
775         775   Peugeot                       207         5490   
131         131    Toyota                     Yaris         5990   
503         503      SEAT                     Altea         5990   
..          ...       ...                       ...          ...   
235         235      Land  Rover Range Rover Evoque        48990   
615         615  Mercedes              Benz Clase C        49990   
37           37       BMW                   Serie 4        57990   
321         321  Mercedes      Benz Clase GLE Coupé        58990   
371         371       BMW                        X5        82580   

                            engine  year  mileage (kms)      fuel     gearbox  \
708                1.25 60cv Urban  2011         116300  Gasolina      Manual   
779                     1.4 HDI SX  2004          58000    Diésel      Manual   
775          Sport 1.6 VTi 16v 120  2007          87450  Gasolina      Manual   
131                     90D ACTIVE  2012         144437    Diésel      Manual   
503  1.6 TDI 105cv Style Ecomotive  2010         132699    Diésel      Manual   
..                             ...   ...            ...       ...         ...   
235            2.0 P250 S AUTO 4WD  2019          33703   Híbrido  Automatica   
615       Mercedes-AMG C 63 Estate  2017          63202  Gasolina  Automatica   
37                  M4 COMPETITION  2018          28999  Gasolina  Automatica   
321     Mercedes-AMG GLE 43 4MATIC  2018         111000  Gasolina  Automatica   
371                      xDrive30d  2021          14000   Híbrido  Automatica   

       location Unnamed: 10  
708  Granollers         NaN  
779  Granollers         NaN  
775  Sabadell 2         NaN  
131    Esplugas         NaN  
503        Gavá         NaN  
..          ...         ...  
235        Gavá         NaN  
615      Mataró         NaN  
37         Gavá         NaN  
321  Granollers         NaN  
371  Sabadell 1         NaN  

[792 rows x 11 columns]

##### Arranged values of data in ascending order by using ' price (eru)' variable.

In [12]:
sample.sort_values(['price (eur)'],ascending=False)

Unnamed: 0     brand                     model  price (eur)  \
371         371       BMW                        X5        82580   
321         321  Mercedes      Benz Clase GLE Coupé        58990   
37           37       BMW                   Serie 4        57990   
615         615  Mercedes              Benz Clase C        49990   
235         235      Land  Rover Range Rover Evoque        48990   
..          ...       ...                       ...          ...   
78           78   Citroen                        C4         5990   
131         131    Toyota                     Yaris         5990   
775         775   Peugeot                       207         5490   
779         779   Citroen                        C2         4990   
708         708      Ford                    Fiesta         4990   

                         engine  year  mileage (kms)      fuel     gearbox  \
371                   xDrive30d  2021          14000   Híbrido  Automatica   
321  Mercedes-AMG GLE 43 4MATIC  2018         111000  Gasolina  Automatica   
37               M4 COMPETITION  2018          28999  Gasolina  Automatica   
615    Mercedes-AMG C 63 Estate  2017          63202  Gasolina  Automatica   
235         2.0 P250 S AUTO 4WD  2019          33703   Híbrido  Automatica   
..                          ...   ...            ...       ...         ...   
78   1.6 e-HDi 110cv Collection  2012         199990    Diésel      Manual   
131                  90D ACTIVE  2012         144437    Diésel      Manual   
775       Sport 1.6 VTi 16v 120  2007          87450  Gasolina      Manual   
779                  1.4 HDI SX  2004          58000    Diésel      Manual   
708             1.25 60cv Urban  2011         116300  Gasolina      Manual   

       location Unnamed: 10  
371  Sabadell 1         NaN  
321  Granollers         NaN  
37         Gavá         NaN  
615      Mataró         NaN  
235        Gavá         NaN  
..          ...         ...  
78     Esplugas         NaN  
131    Esplugas         NaN  
775  Sabadell 2         NaN  
779  Granollers         NaN  
708  Granollers         NaN  

[792 rows x 11 columns]

##### Arranged values of data in descending order by using ' price (eru)' variable.

In [30]:
sample['mileage (kms)'].value_counts()

60000     6
102000    5
110000    4
137000    4
56000     3
         ..
68300     1
99738     1
118069    1
65833     1
9800      1
Name: mileage (kms), Length: 686, dtype: int64

##### mileage (kms) length is 686 and Data type is Int64.

In [14]:
sample.drop_duplicates(inplace=True)

In [15]:
sample

Unnamed: 0       brand     model  price (eur)  \
0             0        SEAT     Ibiza         8990   
1             1     Hyundai       i30         9990   
2             2         BMW   Serie 5        13490   
3             3  Volkswagen      Golf        24990   
4             4        Opel     Corsa        10460   
..          ...         ...       ...          ...   
787         787  Volkswagen      Golf        13990   
788         788         Kia  Sportage        24990   
789         789      Abarth       500        17990   
790         790  Volkswagen    Tiguan        14990   
791         791       Dacia    Dokker        14390   

                                       engine  year  mileage (kms)      fuel  \
0                       SC 1.2 TSI 90cv Style  2016          67000  Gasolina   
1                        1.6 CRDi 110cv Tecno  2014         104868    Diésel   
2                                530d Touring  2011         137566    Diésel   
3                   GTI 2.0 TSI 169kW (230CV)  2018          44495  Gasolina   
4                        1.4 Expression 90 CV  2016          69800  Gasolina   
..                                        ...   ...            ...       ...   
787                 Edition 1.6 TDI 110CV BMT  2016          84040    Diésel   
788            1.6 GDi 97kW (132CV) Basic 4x2  2018          65872  Gasolina   
789  1.4 16v T-Jet 595 118kW (160CV) Pista E6  2019          28830  Gasolina   
790  2.0 TDI 177cv DSG 4x4 Sport BMotion Tech  2014         162895    Diésel   
791            SL Trotamundos dCi 66kW (90CV)  2018           9800    Diésel   

        gearbox    location Unnamed: 10  
0        Manual  Granollers         NaN  
1        Manual  Viladecans         NaN  
2    Automatica  Viladecans              
3        Manual  Viladecans         NaN  
4        Manual  Sabadell 1         NaN  
..          ...         ...         ...  
787      Manual        Gavá         NaN  
788      Manual  Viladecans         NaN  
789      Manual      Mataró         NaN  
790  Automatica      Mataró         NaN  
791      Manual    Badalona         NaN  

[792 rows x 11 columns]

##### In order to Display the duplicates in our dataset.

In [16]:
sample_subset=sample[['brand','model']]

In [17]:
sample_subset

brand     model
0          SEAT     Ibiza
1       Hyundai       i30
2           BMW   Serie 5
3    Volkswagen      Golf
4          Opel     Corsa
..          ...       ...
787  Volkswagen      Golf
788         Kia  Sportage
789      Abarth       500
790  Volkswagen    Tiguan
791       Dacia    Dokker

[792 rows x 2 columns]

##### Coulmn subset is created for 'brand' and 'model'.

In [18]:
sample_row=sample[sample['price (eur)']>20000]

In [19]:
sample_row

Unnamed: 0       brand           model  price (eur)  \
3             3  Volkswagen            Golf        24990   
6             6        Jeep        Wrangler        23990   
9             9        SEAT            Leon        21990   
12           12         BMW         Serie 1        25490   
13           13        SEAT            Leon        23990   
..          ...         ...             ...          ...   
742         742        SEAT           Ateca        24980   
747         747    Mercedes  Benz Clase CLA        22990   
771         771         BMW         Serie 4        32990   
777         777        Jeep        Wrangler        37990   
788         788         Kia        Sportage        24990   

                                      engine  year  mileage (kms)      fuel  \
3                  GTI 2.0 TSI 169kW (230CV)  2018          44495  Gasolina   
6                           3.8 Rubicon Auto  2011         103000  Gasolina   
9    1.5 eTSI 110kW DSG-7 S&S Xcellence Go M  2020          18999   Híbrido   
12                                      118i  2019          27985  Gasolina   
13   2.0 TSI 221kW (300CV) DSG-6 St&Sp CUPRA  2018          40250  Gasolina   
..                                       ...   ...            ...       ...   
742          2.0 TDI 110kW (150CV) S&S Style  2021           4600    Diésel   
747             CLA 220 CDI 4M Aut. AMG Line  2015          65985    Diésel   
771                          420d Automática  2017          50300    Diésel   
777                     2.8 CRD Rubicon Auto  2008         142020    Diésel   
788           1.6 GDi 97kW (132CV) Basic 4x2  2018          65872  Gasolina   

        gearbox    location Unnamed: 10  
3        Manual  Viladecans         NaN  
6    Automatica  Granollers         NaN  
9    Automatica        Gavá         NaN  
12   Automatica      Mataró         NaN  
13   Automatica      Mataró         NaN  
..          ...         ...         ...  
742      Manual    Badalona         NaN  
747  Automatica  Viladecans         NaN  
771  Automatica  Granollers         NaN  
777  Automatica  Sabadell 2         NaN  
788      Manual  Viladecans         NaN  

[144 rows x 11 columns]

##### Row subset is created for vale 'price (eru)' in the source variable.

In [31]:
sample_agg=sample.groupby('price (eur)')['mileage (kms)'].mean()

In [32]:
sample_agg

price (eur)
4990      87150.0
5490      87450.0
5990     159042.0
6490      97806.0
6690     132000.0
           ...   
48990     33703.0
49990     63202.0
57990     28999.0
58990    111000.0
82580     14000.0
Name: mileage (kms), Length: 170, dtype: float64

##### Performed aggregation using groupby.

In [21]:
sample_index=sample.set_index('year')

In [22]:
sample_index

Unnamed: 0       brand     model  price (eur)  \
year                                                  
2016           0        SEAT     Ibiza         8990   
2014           1     Hyundai       i30         9990   
2011           2         BMW   Serie 5        13490   
2018           3  Volkswagen      Golf        24990   
2016           4        Opel     Corsa        10460   
...          ...         ...       ...          ...   
2016         787  Volkswagen      Golf        13990   
2018         788         Kia  Sportage        24990   
2019         789      Abarth       500        17990   
2014         790  Volkswagen    Tiguan        14990   
2018         791       Dacia    Dokker        14390   

                                        engine  mileage (kms)      fuel  \
year                                                                      
2016                     SC 1.2 TSI 90cv Style          67000  Gasolina   
2014                      1.6 CRDi 110cv Tecno         104868    Diésel   
2011                              530d Touring         137566    Diésel   
2018                 GTI 2.0 TSI 169kW (230CV)          44495  Gasolina   
2016                      1.4 Expression 90 CV          69800  Gasolina   
...                                        ...            ...       ...   
2016                 Edition 1.6 TDI 110CV BMT          84040    Diésel   
2018            1.6 GDi 97kW (132CV) Basic 4x2          65872  Gasolina   
2019  1.4 16v T-Jet 595 118kW (160CV) Pista E6          28830  Gasolina   
2014  2.0 TDI 177cv DSG 4x4 Sport BMotion Tech         162895    Diésel   
2018            SL Trotamundos dCi 66kW (90CV)           9800    Diésel   

         gearbox    location Unnamed: 10              cars  
year                                                        
2016      Manual  Granollers         NaN         SEATIbiza  
2014      Manual  Viladecans         NaN        Hyundaii30  
2011  Automatica  Viladecans                    BMWSerie 5  
2018      Manual  Viladecans         NaN    VolkswagenGolf  
2016      Manual  Sabadell 1         NaN         OpelCorsa  
...          ...         ...         ...               ...  
2016      Manual        Gavá         NaN    VolkswagenGolf  
2018      Manual  Viladecans         NaN       KiaSportage  
2019      Manual      Mataró         NaN         Abarth500  
2014  Automatica      Mataró         NaN  VolkswagenTiguan  
2018      Manual    Badalona         NaN       DaciaDokker  

[792 rows x 11 columns]

##### To change the default index position of the 'year' variable.

In [23]:
sample_loc=sample.loc[(sample.year>=2012)&(sample.fuel=='Diésel')]

In [24]:
sample_loc

Unnamed: 0       brand                  model  price (eur)  \
1             1     Hyundai                    i30         9990   
8             8        Land  Rover Discovery Sport        19490   
11           11        Fiat                  Doblò         9990   
19           19  Volkswagen                 Touran        18490   
20           20         BMW                Serie 3        15990   
..          ...         ...                    ...          ...   
771         771         BMW                Serie 4        32990   
785         785     Hyundai                 Tucson        15990   
787         787  Volkswagen                   Golf        13990   
790         790  Volkswagen                 Tiguan        14990   
791         791       Dacia                 Dokker        14390   

                                       engine  year  mileage (kms)    fuel  \
1                        1.6 CRDi 110cv Tecno  2014         104868  Diésel   
8               2.0L eD4 110kW (150CV) 4x2 SE  2017         148575  Diésel   
11     Panorama Lounge 1.6 Multijet 105cv E5+  2015         110642  Diésel   
19               Edition 1.6 TDI CR 115CV BMT  2016          46993  Diésel   
20                          320d Gran Turismo  2014         118200  Diésel   
..                                        ...   ...            ...     ...   
771                           420d Automática  2017          50300  Diésel   
785              1.7 CRDi BlueDrive Tecno 4x2  2015         159416  Diésel   
787                 Edition 1.6 TDI 110CV BMT  2016          84040  Diésel   
790  2.0 TDI 177cv DSG 4x4 Sport BMotion Tech  2014         162895  Diésel   
791            SL Trotamundos dCi 66kW (90CV)  2018           9800  Diésel   

        gearbox    location Unnamed: 10                       cars  
1        Manual  Viladecans         NaN                 Hyundaii30  
8        Manual        Gavá         NaN  LandRover Discovery Sport  
11       Manual      Mataró         NaN                  FiatDoblò  
19       Manual  Sabadell 2         NaN           VolkswagenTouran  
20       Manual  Sabadell 2         NaN                 BMWSerie 3  
..          ...         ...         ...                        ...  
771  Automatica  Granollers         NaN                 BMWSerie 4  
785      Manual        Gavá         NaN              HyundaiTucson  
787      Manual        Gavá         NaN             VolkswagenGolf  
790  Automatica      Mataró         NaN           VolkswagenTiguan  
791      Manual    Badalona         NaN                DaciaDokker  

[353 rows x 12 columns]

##### Specific value 'Diesel' is only located in the data

In [25]:
sample_iloc=sample.iloc[1:3,2:4]

In [26]:
sample_iloc

model  price (eur)
1      i30         9990
2  Serie 5        13490

##### Specific observations and Variables [1:3,2:4] are located in the using Index.


In [27]:
sample_agg=sample.groupby('price (eur)')['year'].mean()

##### function where the values of multiple rows are grouped together to form a single summary value. this dataset can't do any group function.

In [28]:
sample_index=sample.reset_index()

In [29]:
sample_index

index  Unnamed: 0       brand     model  price (eur)  \
0        0           0        SEAT     Ibiza         8990   
1        1           1     Hyundai       i30         9990   
2        2           2         BMW   Serie 5        13490   
3        3           3  Volkswagen      Golf        24990   
4        4           4        Opel     Corsa        10460   
..     ...         ...         ...       ...          ...   
787    787         787  Volkswagen      Golf        13990   
788    788         788         Kia  Sportage        24990   
789    789         789      Abarth       500        17990   
790    790         790  Volkswagen    Tiguan        14990   
791    791         791       Dacia    Dokker        14390   

                                       engine  year  mileage (kms)      fuel  \
0                       SC 1.2 TSI 90cv Style  2016          67000  Gasolina   
1                        1.6 CRDi 110cv Tecno  2014         104868    Diésel   
2                                530d Touring  2011         137566    Diésel   
3                   GTI 2.0 TSI 169kW (230CV)  2018          44495  Gasolina   
4                        1.4 Expression 90 CV  2016          69800  Gasolina   
..                                        ...   ...            ...       ...   
787                 Edition 1.6 TDI 110CV BMT  2016          84040    Diésel   
788            1.6 GDi 97kW (132CV) Basic 4x2  2018          65872  Gasolina   
789  1.4 16v T-Jet 595 118kW (160CV) Pista E6  2019          28830  Gasolina   
790  2.0 TDI 177cv DSG 4x4 Sport BMotion Tech  2014         162895    Diésel   
791            SL Trotamundos dCi 66kW (90CV)  2018           9800    Diésel   

        gearbox    location Unnamed: 10              cars  
0        Manual  Granollers         NaN         SEATIbiza  
1        Manual  Viladecans         NaN        Hyundaii30  
2    Automatica  Viladecans                    BMWSerie 5  
3        Manual  Viladecans         NaN    VolkswagenGolf  
4        Manual  Sabadell 1         NaN         OpelCorsa  
..          ...         ...         ...               ...  
787      Manual        Gavá         NaN    VolkswagenGolf  
788      Manual  Viladecans         NaN       KiaSportage  
789      Manual      Mataró         NaN         Abarth500  
790  Automatica      Mataró         NaN  VolkswagenTiguan  
791      Manual    Badalona         NaN       DaciaDokker  

[792 rows x 13 columns]

##### To reset the index position of 'year' variable we have used reset_index.